### PR TITLE
Avoid enforcing extra-only constraints

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

In the dependency refactor (https://github.com/astral-sh/uv/pull/4430), the logic for requirements and constraints was combined. Specifically, we were applying constraints _before_ filtering on markers and extras, and then applying that same filtering to the constraints. As a result, constraints that should only be activated when an extra is enabled were being enabled unconditionally.

Closes https://github.com/astral-sh/uv/issues/4569.

### CODE REVIEW

File: req.txt [hunk 1]
```diff
@@ -1,3 +0,0 @@
---index-url file:///Users/crmarsh/workspace/packse/index/simple-html/
-
-example-a-961b4c22
```
Comment (charliermarsh) [inline req.txt:3]: Sorry, I must've checked this in at some point accidentally.

File: crates/uv-resolver/src/resolver/mod.rs [hunk 2]
```diff
@@ -1353,8 +1344,58 @@ impl<InstalledPackages: InstalledPackagesProvider> ResolverState<InstalledPackag
...
+                                    ) {
+                                        return false;
+                                    }
+                                }
+                                None => {
+                                    if !constraint.evaluate_markers(self.markers.as_ref(), &[]) {
+                                        return false;
+                                    }
+                                }
+                            }
```
Comment (charliermarsh) [inline crates/uv-resolver/src/resolver/mod.rs:1393]: This is almost the same as above but not quite (the filtering on extras is different), just as it was prior to the refactor (intentionally so).

File: crates/uv-resolver/src/resolver/mod.rs [hunk 3]
```diff
@@ -1302,15 +1302,6 @@ impl<InstalledPackages: InstalledPackagesProvider> ResolverState<InstalledPackag
...
             .apply(dependencies)
-            .flat_map(|requirement| {
-                iter::once(requirement).chain(
-                    // If the requirement was constrained, add those constraints.
-                    self.constraints
-                        .get(&requirement.name)
-                        .into_iter()
-                        .flatten(),
-                )
-            })
```
Comment (charliermarsh) [inline crates/uv-resolver/src/resolver/mod.rs:1313]: The key is that we shouldn't apply the constraints here; we only apply them _after_ we've filtered requirements.

**Comment by charliermarsh** [comment]
Link to the problematic change: https://github.com/astral-sh/uv/pull/4430/files#diff-6be6d80fe4821c47b70a372260f55e73b8da8182b8dcad7525d5cd3eb584532bR1285

**Review by zanieb — APPROVED** [review]
Thanks for pulling up the additional context for me!

**Pull request merged**